Makefile

```diff
@@ -54,12 +56,21 @@ endif
 install:
 	$(PYTHON_INTERPRETER) setup.py install
 
+## Build docker image from Dockerfile (auto-taggged as jason-c-kwan/autometa:<current-branch>)
+image: Dockerfile
+	docker build . -t jason-c-kwan/autometa:`git branch --show-current`
+
 ## Build documentation for autometa.readthedocs.io
 docs:
-	make clean html -C docs && open docs/build/html/index.html
+	make clean html -C docs
+	@echo "docs built. Open docs/build/html/index.html to view"
+
+## Download test_data.json for unit testing
+unit_test_data_download:
+	gdown --id $(TEST_DATA_FILEID) -O tests/data/test_data.json
 
-## Build test_data.json file for unit testing
-unit_test_data: tests/data/
+## Build test_data.json file for unit testing (requires all files from https://drive.google.com/open?id=189C6do0Xw-X813gspsafR9r8m-YfbhTS be downloaded into tests/data/)
+unit_test_data_build: tests/data/records.fna
```

Comment:
So I tried `make image` which worked on my mac, then in the image, after doing `apt-get install build-essential` ran `make unit_test_data_download` (which worked), I tried `make unit_test_data_build`, and I got this:

```
make: *** No rule to make target 'tests/data/records.fna', needed by 'unit_test_data_build'.  Stop.
```

Reply:
________________________________________________________________________

Did you try `make unit_test`?

`make unit_test_data_build` is the first command to use if you were trying to rebuild the `test_data.json` object. This will aggregate all of the files found in the path linked to google drive in the docs. 

________________________________________________________________________
Perhaps for documentation
@Sidduppal 
________________________________________________________________________

# Building Autometa's unit tests

## Relevant Makefile Commands

- `unit_test_data_build`
- `unit_test_data_download`
- `unit_test_wip`
- `unit_test_entrypoints`
- `unit_test`

### `unit_test_data_build`

This command is used to build the `test_data.json` file for unit testing. I.e. it will run the script `make_test_data.py` which will aggregate all of the files in the `tests/data` folder (Note: the required files have been placed here https://drive.google.com/open?id=189C6do0Xw-X813gspsafR9r8m-YfbhTS and should be downloaded into `tests/data/`). This is the first or perhaps 0th step when it comes to running the tests as it generates the `test_data.json` file that is parsed to retrieve all of the pre-generated variables used for intermediate stages of the pipeline. This is done to reduce the test time and computational workload when running through the test suite.

As an example, running diamond BLAST is a stage in the pipeline, but we _should not_ be testing whether diamond works, so we simply need to retrieve a "pre-diamond blasted table" and ensure our parser as well as subsequent steps are functioning appropriately. Therefore, the table is written into `test_data.json` and retrieved when the parser unit test is conducted. Decisions at this stage are somewhat fundamental and may have unintended side-effects across the rest of the test suite.

### `unit_test_data_download`

This command downloads the `test_data.json` object that was generated from `unit_test_data_build` and then uploaded to the autometa_test_data google drive. This provides convenience for someone wanting to run the test suite that requires the test variables constructed from the previous command. This is a necessary step when wanting to run unit tests as the `test_data.json` file will hold many of the variables necessary to conduct these tests.

## `unit_test_wip`

This command runs unit tests but only includes tests marked as work-in-progress (WIP). This is denoted in pytest with the decorator:

```python
@pytest.mark.wip
def test_some_function(...):
    ...
```

## `unit_test_entrypoints`

This command runs unit tests but only includes tests marked as entrypoints. This is denoted in pytest with the decorator:

```python
@pytest.mark.entrypoint
def test_some_function_that_is_an_entrypoint(...):
    ...
```

Entrypoints are basically the commands we will list out as 'console scripts' in `setup.py` that will be available as commands to the end user. These are important and sometimes referred to as "happy" tests because if one of these fail for the end-user, they will probably be quite unhappy and likely distrust the functionality of the rest of the codebase.

## `unit_test`

This command runs all unit tests under the `tests` directory. This includes all tests marked as WIP or as entrypoints. _However_ this _will_ skip tests marked with the decorator:

```python
@pytest.mark.skip
def test_some_function(...):
    ...
```

Reply:
From the above description, it is unclear whether you need to run `make unit_test_data_build` as well as `make unit_test_data_download`. I just pulled this PR, ran `make image`, then in the image successfully ran `make unit_test_data_download`. After that, `make unit_test_data_build` gives me:

```
make: *** No rule to make target 'tests/data/records.fna', needed by 'unit_test_data_build'.  Stop.
```

`make unit_test_wip` gives me:

```
python3 -m pytest -m "wip" --durations=0 --cov=autometa --emoji --cov-report html
ERROR: usage: __main__.py [options] [file_or_dir] [file_or_dir] [...]
__main__.py: error: unrecognized arguments: --variables --cov=autometa --emoji --cov-report html
  inifile: /pytest.ini
  rootdir: /

make: *** [Makefile:85: unit_test_wip] Error 4
```

Then `make unit_test_entrypoints` gives me:

```
python3 -m pytest -m "entrypoint" --durations=0 --cov=autometa --emoji --cov-report html
ERROR: usage: __main__.py [options] [file_or_dir] [file_or_dir] [...]
__main__.py: error: unrecognized arguments: --variables --cov=autometa --emoji --cov-report html
  inifile: /pytest.ini
  rootdir: /

make: *** [Makefile:89: unit_test_entrypoints] Error 4
```

I also get the same thing with `make unit_test`. Seems like there is something wrong with `__main__.py`? Not sure what the `make unit_test_data_build` error is all about.

Reply:
`pytest-variables` and `pytest-cov` need to be installed. I’ve simplified this a bit more within the Makefile by adding the command: `test_environment`. Now if you run `make unit_test` (or any of these that have `test_environment` as a dependency) it will install any requirements if they have not been installed prior to running the unit tests

Reply:
OK, I re-pulled, and then deleted my old docker image and re-built it. I ran `make test_environment`, which seemed to work fine, then `make unit_test_data_download`, which also worked. But then when I tried `make unit_test_data_build`, I still get:

```
make: *** No rule to make target 'tests/data/records.fna', needed by 'unit_test_data_build'.  Stop.
```

I now get a bit further with `make unit_test_wip`. I get:

```
...
============================================================================================================ test session starts =============================================================================================================
platform linux -- Python 3.8.5, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /, configfile: pytest.ini
plugins: html-3.1.1, metadata-1.11.0, md-0.2.0, emoji-0.2.0, repeat-0.9.1, cov-2.11.1, variables-1.9.0
collected 0 items / 1 error

=================================================================================================================== ERRORS ===================================================================================================================
_______________________________________________________________________________________________________ ERROR collecting test session ________________________________________________________________________________________________________
opt/conda/lib/python3.8/site-packages/_pytest/runner.py:311: in from_call
    result: Optional[TResult] = func()
opt/conda/lib/python3.8/site-packages/_pytest/runner.py:341: in <lambda>
    call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
opt/conda/lib/python3.8/site-packages/_pytest/main.py:710: in collect
    for x in self._collectfile(path):
opt/conda/lib/python3.8/site-packages/_pytest/main.py:546: in _collectfile
    assert (
E   AssertionError: local('/dev/console') is not a file (isdir=False, exists=True, islink=False)

----------- coverage: platform linux, python 3.8.5-final-0 -----------
Coverage HTML written to dir htmlcov

========================================================================================================== short test summary info ===========================================================================================================
ERROR  - AssertionError: local('/dev/console') is not a file (isdir=False, exists=True, islink=False)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================================================== 1 error in 0.82s ==============================================================================================================
make: *** [Makefile:89: unit_test_wip] Error 2
```

I get a similar thing for `make unit_test_entrypoints`:

```
============================================================================================================ test session starts =============================================================================================================
platform linux -- Python 3.8.5, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /, configfile: pytest.ini
plugins: html-3.1.1, metadata-1.11.0, md-0.2.0, emoji-0.2.0, repeat-0.9.1, cov-2.11.1, variables-1.9.0
collected 0 items / 1 error

=================================================================================================================== ERRORS ===================================================================================================================
_______________________________________________________________________________________________________ ERROR collecting test session ________________________________________________________________________________________________________
opt/conda/lib/python3.8/site-packages/_pytest/runner.py:311: in from_call
    result: Optional[TResult] = func()
opt/conda/lib/python3.8/site-packages/_pytest/runner.py:341: in <lambda>
    call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
opt/conda/lib/python3.8/site-packages/_pytest/main.py:710: in collect
    for x in self._collectfile(path):
opt/conda/lib/python3.8/site-packages/_pytest/main.py:546: in _collectfile
    assert (
E   AssertionError: local('/dev/console') is not a file (isdir=False, exists=True, islink=False)

----------- coverage: platform linux, python 3.8.5-final-0 -----------
Coverage HTML written to dir htmlcov

========================================================================================================== short test summary info ===========================================================================================================
ERROR  - AssertionError: local('/dev/console') is not a file (isdir=False, exists=True, islink=False)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================================================== 1 error in 0.50s ==============================================================================================================
make: *** [Makefile:93: unit_test_entrypoints] Error 2
```

This StackOverflow seems to suggest the last two problems are something to do with setting the working directory correctly. Really not sure what is going on with `make unit_test_data_build`.

https://stackoverflow.com/questions/65394782/pytest-not-working-with-django-and-docker-assertionerror-local-dev-console

Reply:
## Addressing `unit_test_data_build`

`unit_test_data_build` is **_not_** required for running unit tests. This is a one-time command if one plans to create a new `test_data.json` object. This would only be necessary if we are changing file formats or adding more objects into the test suite.

The only commands you need to run the test suite are `unit_test_data_download` and `unit_test{_wip,_entrypoints}`.

You will continue to get the error:

```
make: *** No rule to make target 'tests/data/records.fna', needed by 'unit_test_data_build'.  Stop.
```
 
If you do not download the entire directory within [autometa_test_data/unit_test_data](https://drive.google.com/drive/folders/189C6do0Xw-X813gspsafR9r8m-YfbhTS) that contains [records.fna](https://drive.google.com/file/d/1DP7U7Lno8VW8Kn5Y2tjK3yP6pGnrMP1P/view?usp=sharing) as well as all of the other files that were used to create test_data.json.

Make is telling you that the file at the path `tests/data/records.fna` does not exists. This file is not downloaded with `unit_test_data_download`. This file, as well as a number of other files in [autometa_test_data/unit_test_data](https://drive.google.com/drive/folders/189C6do0Xw-X813gspsafR9r8m-YfbhTS) would need to be downloaded for `unit_test_data_build` to function appropriately.

Luckily, the user only needs the `test_data.json` file to run tests, so you can safely ignore the above command when performing unit tests for autometa

Reply:
OK, I was able to run the tests using these instructions (and the last commit). Thanks!